resource/config_map.go

```diff
@@ -0,0 +1,259 @@
+
+		desired, err = c.desiredFunc(ctx, obj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// Enrich desired objects with labels to be able to select and
+		// clean them during the deletion.
+		for _, d := range desired {
+			err = setLabels(&d, obj, c.Name())
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "found desired ConfigMap objects")
+	}
+
+	for _, d := range desired {
+		err := c.ensureCreated(ctx, &d, obj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (c *ConfigMap) ensureCreated(ctx context.Context, desired *corev1.ConfigMap, obj interface{}) error {
+	var err error
+
+	printName := newPrintName(desired)
+
+	var current *corev1.ConfigMap
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding current ConfigMap %#q", printName))
+
+		opts := metav1.GetOptions{
+			IncludeUninitialized: true,
+		}
+
+		current, err = c.k8sClient.CoreV1().ConfigMaps(desired.Namespace).Get(desired.Name, opts)
+		if apierrors.IsNotFound(err) {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find current ConfigMap %#q", printName))
+
+		} else if err != nil {
+			return microerror.Mask(err)
+
+		} else {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found current ConfigMap %#q", printName))
+		}
+	}
+
+	if current != nil {
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("updating ConfigMap %#q", printName))
+
+		// Make sure the object is managed by this resource by checking
+		// labels.
+		err = assertLabels(current, obj, c.Name())
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		toUpdate := newConfigMapToUpdate(current, desired)
+
+		if toUpdate != nil {
+			_, err := c.k8sClient.CoreV1().ConfigMaps(toUpdate.Namespace).Update(toUpdate)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("updated ConfigMap %#q", printName))
+
+		} else {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("ConfigMap %#q is up to date", printName))
+		}
+
+	} else {
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("creating ConfigMap %#q", printName))
+
+		_, err := c.k8sClient.CoreV1().ConfigMaps(desired.Namespace).Create(desired)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("created ConfigMap %#q", printName))
+	}
+
+	return nil
+}
+
+func (c *ConfigMap) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	var err error
+
+	var labelSelector string
+	{
+		c.logger.LogCtx(ctx, "level", "info", "message", "creating label selector")
+
+		labelSelector, err = newLabelSelector(obj, c.Name())
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "info", "message", "created label selector")
+	}
```

Comment:
We use label selector to delete objects created by this resource for the `obj`. That way there is no need to call `DesiredFunc` during deletion or passing `Name` and `Namespace` with the `Config` struct.